File: Unity/Assets/PixelPlanets/StellarBreaker/Scripts/Economy/CurrencyService.cs
using System;
using StellarBreaker.Core;

namespace StellarBreaker.Economy
{
    /// <summary>
    /// Holds the player's Stardust balance as a BigNumber. Scene-independent.
    /// </summary>
    public class CurrencyService
    {
        public BigNumber Stardust { get; private set; } = BigNumber.Zero;

        /// <summary>Fired whenever the balance changes; passes the new balance.</summary>
        public event Action<BigNumber> OnChanged;

        public void Add(BigNumber amount)
        {
            if (amount <= BigNumber.Zero) return;
            Stardust = Stardust + amount;
            OnChanged?.Invoke(Stardust);
        }

        public bool CanAfford(BigNumber cost) => Stardust >= cost;

        /// <summary>Deduct cost if affordable. Returns false (no change) otherwise.</summary>
        public bool TrySpend(BigNumber cost)
        {
            if (cost < BigNumber.Zero) return false;
            if (Stardust < cost) return false;
            Stardust = Stardust - cost;
            OnChanged?.Invoke(Stardust);
            return true;
        }

        /// <summary>Directly set the balance (for save/load or tests).</summary>
        public void Set(BigNumber value)
        {
            Stardust = value;
            OnChanged?.Invoke(Stardust);
        }
    }
}

File: Telegram/TelegramApp/src/game/economy/CurrencyService.ts
// Ported from Assets/PixelPlanets/StellarBreaker/Scripts/Economy/CurrencyService.cs
import { BigNumber } from '../core/BigNumber'
import { Emitter } from '../core/Emitter'

/** Holds a currency balance (Stardust or Relics) as a BigNumber. */
export class CurrencyService {
  private _balance: BigNumber = BigNumber.Zero

  /** Fired whenever the balance changes; passes the new balance. */
  readonly onChanged = new Emitter<BigNumber>()

  get balance(): BigNumber {
    return this._balance
  }

  add(amount: BigNumber): void {
    if (amount.lte(BigNumber.Zero)) return
    this._balance = this._balance.add(amount)
    this.onChanged.emit(this._balance)
  }

  canAfford(cost: BigNumber): boolean {
    return this._balance.gte(cost)
  }

  /** Deduct cost if affordable. Returns false (no change) otherwise. */
  trySpend(cost: BigNumber): boolean {
    if (cost.lt(BigNumber.Zero)) return false
    if (this._balance.lt(cost)) return false
    this._balance = this._balance.sub(cost)
    this.onChanged.emit(this._balance)
    return true
  }

  /** Directly set the balance (for save/load or tests). */
  set(value: BigNumber): void {
    this._balance = value
    this.onChanged.emit(this._balance)
  }
}
